Provide public interface for skipping inline page images

## Explanation

I want to extract all images from a page, but omit inline images as they are not really useful in my case and just generate overhead (2 ms without and 29 s with inline images for one page with a dotted table which has 24643 inline images, but no "real" images).

## Code Example

For now, I am basically exploiting https://github.com/py-pdf/pypdf/blob/c2a741e968da1b4d7195a7924ba62a9445f825e4/pypdf/_page.py#L478 which does not seem to be a clean solution:

```python
from pypdf import PdfReader

reader = PdfReader(path)
for page in reader.pages:
    page.inline_images = dict()  # Avoid loading inline images.
    for image in page.images:
        print(image)
```

@pubpub-zz I just tested your solution. While it works on the example file and I can see the desired speed improvements, the filtering fails when the key is a list (which is easy to mitigate on my side, but I just wanted to point this out):

```python3
[...]
/Image185
/Image185 ImageFile(name=Image185.png, data: 525 Byte)
['/Meta187', '/Image188']
Traceback (most recent call last):
  File "/home/stefan/temp/test.py", line 8, in <module>
    if key.startswith('~'):
AttributeError: 'list' object has no attribute 'startswith'
```

This might be similar to what mypy complains about with

> pypdf/_page.py:505: error: Unsupported operand types for + ("List[Union[str, List[str]]]" and "List[str]")  [operator]

as well.

> @pubpub-zz I just tested your solution. While it works on the example file and I can see the desired speed improvements, 
the filtering fails when the key is a list (which is easy to mitigate on my side, but I just wanted to point this out):

I forgot the list output (when images are within XImage
Can you provide your solution for others ?

> This might be similar to what mypy complains about with
> 
> > pypdf/_page.py:505: error: Unsupported operand types for + ("List[Union[str, List[str]]]" and "List[str]")  [operator]

As said in the PR, the mypy issue is under analysis

--- a/pypdf/_page.py
+++ b/pypdf/_page.py
@@ -344,7 +344,8 @@ class PageObject(DictionaryObject):
         DictionaryObject.__init__(self)
         self.pdf: Union[None, PdfReaderProtocol, PdfWriterProtocol] = pdf
         self.inline_images: Optional[Dict[str, ImageFile]] = None
-        self.inline_images_keys: Optional[List[str]] = None
+        # below Union for mypy but actually Optional[List[str]]
+        self.inline_images_keys: Optional[List[Union[str, List[str]]]] = None
         if indirect_ref is not None:  # deprecated
             warnings.warn(
                 (

--- a/pypdf/_page.py
+++ b/pypdf/_page.py
@@ -28,6 +28,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 
 import math
+import re
 import warnings
 from decimal import Decimal
 from typing import (
@@ -55,6 +56,7 @@ from ._text_extraction import (
     mult,
 )
 from ._utils import (
+    WHITESPACES_AS_REGEXP,
     CompressedTransformationMatrix,
     File,
     ImageFile,
@@ -342,6 +344,7 @@ class PageObject(DictionaryObject):
         DictionaryObject.__init__(self)
         self.pdf: Union[None, PdfReaderProtocol, PdfWriterProtocol] = pdf
         self.inline_images: Optional[Dict[str, ImageFile]] = None
+        self.inline_images_keys: Optional[List[str]] = None
         if indirect_ref is not None:  # deprecated
             warnings.warn(
                 (
@@ -475,8 +478,14 @@ class PageObject(DictionaryObject):
     def _get_ids_image(
         self, obj: Optional[DictionaryObject] = None, ancest: Optional[List[str]] = None
     ) -> List[Union[str, List[str]]]:
-        if self.inline_images is None:
-            self.inline_images = self._get_inline_images()
+        if self.inline_images_keys is None:
+            nb_inlines = len(
+                re.findall(
+                    WHITESPACES_AS_REGEXP + b"BI" + WHITESPACES_AS_REGEXP,
+                    self._get_contents_as_bytes() or b"",
+                )
+            )
+            self.inline_images_keys = [f"~{x}~" for x in range(nb_inlines)]
         if obj is None:
             obj = self
         if ancest is None:
@@ -485,7 +494,7 @@ class PageObject(DictionaryObject):
         if PG.RESOURCES not in obj or RES.XOBJECT not in cast(
             DictionaryObject, obj[PG.RESOURCES]
         ):
-            return list(self.inline_images.keys())
+            return self.inline_images_keys
 
         x_object = obj[PG.RESOURCES][RES.XOBJECT].get_object()  # type: ignore
         for o in x_object:
@@ -493,7 +502,7 @@ class PageObject(DictionaryObject):
                 lst.append(o if len(ancest) == 0 else ancest + [o])
             else:  # is a form with possible images inside
                 lst.extend(self._get_ids_image(x_object[o], ancest + [o]))
-        return lst + list(self.inline_images.keys())
+        return lst + self.inline_images_keys
 
     def _get_image(
         self,
@@ -515,6 +524,8 @@ class PageObject(DictionaryObject):
                 raise
         if isinstance(id, str):
             if id[0] == "~" and id[-1] == "~":
+                if self.inline_images is None:
+                    self.inline_images = self._get_inline_images()
                 if self.inline_images is None:  # pragma: no cover
                     raise KeyError("no inline image can be found")
                 return self.inline_images[id]
@@ -893,6 +904,23 @@ class PageObject(DictionaryObject):
             ],
         )
         return contents
+
+    def _get_contents_as_bytes(self) -> Optional[bytes]:
+        """
+        Return the page contents as bytes .
+
+        Returns:
+            The ``/Contents`` object as bytes, or ``None`` if it doesn't exist.
+
+        """
+        if PG.CONTENTS in self:
+            obj = self[PG.CONTENTS].get_object()
+            if isinstance(obj, list):
+                return b"".join(x.get_object().get_data() for x in obj)
+            else:
+                return cast(bytes, cast(EncodedStreamObject, obj).get_data())
+        else:
+            return None
 
     def get_contents(self) -> Optional[ContentStream]:
         """

--- a/pypdf/_page.py
+++ b/pypdf/_page.py
@@ -908,7 +908,7 @@ class PageObject(DictionaryObject):
 
     def _get_contents_as_bytes(self) -> Optional[bytes]:
         """
-        Return the page contents as bytes .
+        Return the page contents as bytes.
 
         Returns:
             The ``/Contents`` object as bytes, or ``None`` if it doesn't exist.

--- a/pypdf/_utils.py
+++ b/pypdf/_utils.py
@@ -382,6 +382,7 @@ def ord_(b: Union[int, str, bytes]) -> Union[int, bytes]:
 
 
 WHITESPACES = (b" ", b"\n", b"\r", b"\t", b"\x00")
+WHITESPACES_AS_REGEXP = b"[ \n\r\t\x00]"
 
 
 def paeth_predictor(left: int, up: int, up_left: int) -> int:

--- a/tests/test_workflows.py
+++ b/tests/test_workflows.py
@@ -1016,9 +1016,6 @@ def test_inline_images():
         _a[x] = y
     with pytest.raises(KeyError) as exc:
         reader.pages[2]._get_image(("test",))
-    reader.pages[2].inline_images = None
-    with pytest.raises(KeyError) as exc:
-        reader.pages[2]._get_image(("~1~",))
 
 
 @pytest.mark.enable_socket()

